Add backup compression

cblanken opened the pull request:
This PR addresses issue #407 

Database backups for LC can be quite large. This adds compression into the `.zip` format with the default compression level for all database backups.

It's pretty simple, just pipes the `mysqldump` through `zip` if it's enabled. I added an option in the `settings` table and seeded it to be enabled by default.

@simjanos-dev 
I did have one question though. Where would you prefer that I place the option to enable/disable the compression. Just on the Dashboard settings page next to the manual "Create backup"? It doesn't seem like it would fit anywhere else, but I wasn't sure if you wanted to keep the "Dashboard" clean of settings and only use it for manual interactions.

cblanken committed:
feat: add database backup compression

Database backups for LC can be quite large. This adds compression into the
`.zip` format with the default compression level for all database backups.

simjanos-dev commented:
> Where would you prefer that I place the option to enable/disable the compression

I think it's a good place for it. I want to add other things to dashboard, like indicators of whether external services are working or not and importing speed statistics. When I added the backup, I think I didn't want to add a separate tab for it.

Feel free to move it around or modify it however you like.

It might take a few days for me to look at this PR. Please also make sure that it works even if people already have backup files when updating to this version.

cblanken committed:
feat: add backup compression UI toggle to settings

Adds "Backups" tab to the admin settings for backup related settings including
the new backup compression option.

cblanken commented:
I added the enable/disable for backup compression to a separate tab in AdminSettings since it occurred to me it would be nice to have a few other settings as well.

@simjanos-dev 
If you aren't opposed I'd like to add these settings:
1. Set the retention count essentially replacing the daily `MAX_SAVED_BACKUPS` env variable.
2. Options for keeping backups at different intervals e.g. weekly, monthly, yearly.
3. Set the time of day for backups.

simjanos-dev commented:
> If you aren't opposed I'd like to add these settings:

It seems like basically moving the settings to the UI itself(2. and 3. can be set with the cron interval config in the docker-compose file). Sounds good to me.

cblanken committed:
feat: add backup scheduling to admin settings

- Enables users to change the backup schedule via Cron expression.
- A Regex (see
  https://gist.github.com/Aterfax/401875eb3d45c9c114bbef69364dd045) is used for
  validation on the frontend to reduce user input error, but more robust
  checking is used before updating the database or schedule.
- If an invalid Cron expression is provided for the `BACKUP_INTERVAL` env var,
  then a default of `0,30 * * * *` is used instead as the default.

cblanken committed:
feat: add backup retention options

Prunes (deletes) old backups based on retention settings for days, weeks,
months, years, and the most recently made backups for better coverage in the
event of data loss.

cblanken committed:
feat: split backup settings tab into sections

Improves the readability of the backup settings and leaves the UI open to
additional backup related options in the future.

cblanken committed:
fix: backup settings fixups after controller refactor

Some of the changes for the contorller refactor broke saving for the new backup
settings.

cblanken commented:
Okay, I think this PR is finally ready for review. I still need to write some tests for the retention rules to see if there are any weird edge cases, but everything is there that I wanted to add.

<img width="774" height="896" alt="image" src="https://github.com/user-attachments/assets/d9261fce-9952-4c5a-83bf-c26d022543b1" />

simjanos-dev commented:
Thank you so much, this looks great!


I'll probably look at it on the weekend or next week.

simjanos-dev commented:
Thanks again for the PR! 

I've tried it out, and the new UI looks great. After trying out the "Options for keeping backups at different intervals e.g. weekly, monthly, yearly.", I feel like it's a bit more complex thank I've expected. Would it be okay to revert it back and only keeping the "most recent" option?


The global settings validation is a great idea. I've made a github issue, I'll add them for the other settings as well.

cblanken commented:
You can if you'd like, but I'd ask you to consider it. The main reason I wanted to add the extra retention levels is because with the existing backup solution of "most recent" backups you lose the opportunity to recover data very quickly. If you have a backup cron schedule of `0 15 30 45 * * * *` which was the default, if you have a maximum recent backup count of 8. You'd only be able to recover data from a maximum of two hours ago. That's not ideal in case there is some data corruption or deletion that happens and isn't immediately identified. The other retention intervals let you keep older backups without filling up your disk unnecessarily.

Personally, I have an external backup solution that I could recover data from, but I doubt everyone does. By default the extra retention options would:
1. Improve backup coverage over longer periods of time
2. Improve the likelihood of successfully recover in case of data loss
3. Reduce disk usage


If you'd like I  could try to redesign a bit to make it more user friendly? Here are some examples of how it might look like, at from how I've seen it implemented elsewhere.
<img width="782" height="401" alt="image" src="https://github.com/user-attachments/assets/fc2305d7-a83b-459f-8980-12e343c6511b" />
<img width="559" height="482" alt="image" src="https://github.com/user-attachments/assets/5f6d99c1-6c82-4064-bae6-645ffb04a8d2" />


Ideally though, these are the kinds of settings that most user's shouldn't need to change anyway. With some reasonable defaults, they'll just improve the chances of recover from data loss.

simjanos-dev commented:
> If you have a backup cron schedule of 0 15 30 45 * * * * which was the default, if you have a maximum recent backup count of 8

On prod it is ` 59 23 * * *`. Which is also an issue, because a lot of users don't run their PC at that time.

> Ideally though, these are the kinds of settings that most user's shouldn't need to change anyway. With some reasonable defaults, they'll just improve the chances of recover from data loss.

That is a good point, I do not want users to lose data. But I would still like to simplify it somehow, the code is also very difficult to follow (not because it is badly written, just the problem is complex).

Maybe we could do something like:
- hourly
- daily
- weekly
- monthly
- yearly

in different folders, and letting users decide with checkboxes if they want these options and removing CRON. It would be simple to check if the current year/month/week/day/hour has a backup or not. But this would cause some duplicate backups if multiple are selected.

To avoid duplicates with this system, we could start checking if a new backup is required by the largest interval, and only creating 1 backup, so If a new yearly is needed, we won't also make a monthly. This would also fix the ` 59 23 * * *` issue mentioned above.

I don't know what to do, I thought about it too much. I'll come back to it a bit later.

cblanken commented:
> That is a good point, I do not want users to lose data. But I would still like to simplify it somehow, the code is also very difficult to follow (not because it is badly written, just the problem is complex).

It's probably at least little bit poorly written 😅 
I'll see if I can add some more comments and maybe make it more succinct. I just couldn't think of a better way to implement it at the time.

cblanken commented:
> Maybe we could do something like:
>
> ...
>
> in different folders, and letting users decide with checkboxes if they want these options and removing CRON. It would be simple to check if the current year/month/week/day/hour has a backup or not. But this would cause some duplicate backups if multiple are selected.

I'm not sure I understand how that would be different from the user's perspective than what's already there? Also how would the user configure the time of day and/or frequency of backups? Wouldn't that have essentially the same issue as the current default `59 23 * * *` schedule?

cblanken committed:
refactor: fixup backup options

- clean up backup option UI
- clean up retention implementation and add comments

cblanken committed:
test: add backup retention baseline test

This test generates a bunch of dummy files (every hour over the maximum years
given in the the options) and executes the backup pruning according to the
default values seeded in the backup settings.

cblanken commented:
Okay, I tried to clean up a little and add some comments. I also added a baseline test for pruning old backups according to the retention settings.

I'm still struggling to come up with a simpler implementation of the pruning process though.

simjanos-dev commented:
> Okay, I tried to clean up a little and add some comments.

I'll try to clean it up a bit as well. I have an hour~, I'll push it to this PR, then I'll leave a few comments. (in case you are still working on it)

> I'm still struggling to come up with a simpler implementation of the pruning process though.

Same. It's fine, it is not that long and separated. There's nothing wrong with it, just complex. But like you said, users probably won't change these settings.

(Sorry for my English, I can't quite express myself in the tone I want. I really appreciate you working on this.)

simjanos-dev commented:
I don't know what happens when `Schedule::command('app:create-backup')->cron($cron)->withoutOverlapping();` is being called elsewher, but as far as I understand, this should only be called once in `routes/console.php`, and shouldn't be used in other parts of the code. 

Also the content of `routes/console.php` is being run each time an artisan command is being run. It will always attempt to make a db request, and possibly runs the `Schedule::command` command. Don't know what happens when it is being called elsewhere, but if it would appear multiple times in `routes/console.php`, then the backup command would be scheduled multiple times, and it would run multiple times. I think it is not documented in laravel what happens if we call it it somewhere like in settingscontroller. 

Now thinking about it, I actually don't know how we could request the cron interval from the user on the UI. Anything I can think of is probably not a recommended way of using the scheduler. 

A few other things:
- `php artisan test --filter=BackupPruning` does not find the test, unless its named BackupPruningTest.php.
- Please don't use public functions for services. The services should be created in the constructors of classes.
- (int) Setting::where('name', 'backupRetainYearly')->first()->value. Setting models have a ->decode() method. But I also just realized, that I think decimals are not being converted to php floats, If I'm correct floats are strings in json, added a todo, I'll check it later.

The code probably can't be simplified, but when there are multiple nested loops:

```
foreach ($retentionIntervals as $retentionInterval) {
          $filesPartitionedByInterval = [];
          $partitionStart = new DateTime('now');
          $partitionEnd = clone $partitionStart;
          $keepCount = $retentionInterval['max'];
          for ($i = 0; $i < $keepCount; $i++) {
              $partitionStart->sub($retentionInterval['interval']);
              $filesPartitionedByInterval[] = array_filter($filesToKeep, fn ($file) => $file['mtime'] > $partitionStart->getTimeStamp() && $file['mtime'] < $partitionEnd->getTimeStamp());
              $partitionEnd = clone $partitionStart;
          }

          // Mark the first (most recent backup) of each interval to be kept
          foreach ($filesPartitionedByInterval as $partition) {
              $partitionKeys = array_keys($partition);
              if (array_key_exists(0, $partitionKeys)) {
                  $firstKey = $partitionKeys[0];
                  $filesToKeep[$firstKey]['keep'] = true;
              }
          }
      }
```

What I usually do is I take each loop, and move them into their function, which would make them more easy to read.

Same with complex lines like this:
```
$filesToKeep = array_replace($filesToKeep, array_map($markKeep, array_slice($filesToKeep, 0, $mostRecentRetentionCount, preserve_keys: true)));
```

Like:
```
$filesToKeep = $this->filterRecentFiles(...);
```

I have to go to sleep. I won't have much time in the next few days, I'll probably be working on refactoring my code if I'll have some.

cblanken commented:
Sorry, didn't mean to make a bunch of work for you. I can fixup all those things you mentioned in your comments if you'd like. And dig into the scheduling a bit more to see if there's a proper way to reconfigure it at runtime. I must admit I'm still a bit of a Laravel newbie 😅 but I'll have some time over the next few days.

Edit:
I found this project (https://github.com/bedezign/laravel-dynamic-scheduler) which is old, but does make me believe it's possible, even if it requires setting up some kind of wrapper to the Scheduler like it mentions in the README.

simjanos-dev commented:
> Sorry, didn't mean to make a bunch of work for you.

It's fine, Its just more complicated than I thought it would be. 

> must admit I'm still a bit of a Laravel newbie

I'm not very familiar with these parts either. 

> I found this project

I try to avoid adding any custom changes like that. Ill check out that project later.

cblanken commented:
> Now thinking about it, I actually don't know how we could request the cron interval from the user on the UI. Anything I can think of is probably not a recommended way of using the scheduler.

and

> I try to avoid adding any custom changes like that. Ill check out that project later.

@simjanos-dev 
Would you like me to revert the CRON scheduling UI changes for the time being? It seems like it might be pretty involved or at least non-standard and may be better to add in a separate PR (if at all).

That way we can move forward with the compression and retention settings. With the retention settings in place, I think it would be reasonable to configure the default backup schedule to be hourly or even every 30 minutes. Which would alleviate the `59 23 * * *` default schedule issue and mostly obviate the need for CRON scheduling in the UI anyway.

cblanken committed:
Revert "feat: add backup scheduling to admin settings"

Modifying a task schedule at runtime in Laravel seems to run against the
typical use case for the Task Scheduler and does not function correctly as it's
currently implemented, so the UI backup schedule configuration via CRON is
being removed for the time being.

This reverts commit 470acf665c9c347222f567f8d7456cf2acd743cb.

cblanken committed:
refactor: backup retention

- Removes `static` accesibility of functions in BackupService
- Refactors `deleteOldBackups` to be more readable
- Refactors complex loops, filters, and maps into their own functions
- Use `Setting->decode()` instead of (int) type conversion
- Fix BackupPruning test name

cblanken commented:
Okay, I reverted the CRON scheduling stuff until we can figure out where to go with it or if it's needed at all.

@simjanos-dev 
I also refactored everything according to your comments above as well as I could. Please let me know if there's anything else you'd like me to address.

cblanken commented:
Screenshot of the finalized backup settings page
<img width="986" height="657" alt="image" src="https://github.com/user-attachments/assets/0f73935f-821c-4628-b091-c00a0ae1846a" />

simjanos-dev commented:
Thank you so much!

Was this tested for using it together with older already existing backups?

simjanos-dev committed:
fix: mysqldump parameter

simjanos-dev committed:
Fixed filename inside the .zip file

simjanos-dev commented:
I added --skip-ssl flag because for some reason mysqldump stopped working withoutit, even on dev branch. I also replaced the zip bash command with gzip, this way the .sql file keeps its name inside the zip file, instead of being named '-'.

Please in the future use: 
- Collections instead of arrays (where it makes sense/possible)
- Laravel Arr:: helpers
- Laravel Str:: helpers
- Carbon and CarbonInterval for dates/times

I also wrote a lot of old code where I didn't use these, I'll try to refactor those whenever I can.
